### ISSUE TITLE
Race condition in Dokan FUSE

### ISSUE THREAD

**Issue by mistersandman** [dokan-dev/dokany#354]
# Environment
 * Windows version: Any
 * Processor architecture: Any
 * Dokany version: latest master (https://github.com/dokan-dev/dokany/commit/3361dbf7e681d45d71753b2931967e0bd700e492)
 * Library type (Dokany/FUSE): FUSE
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [ ] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [ ] I can always reproduce the issue with the provided description below.
  - [ ] I have updated Dokany to the latest version (https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.

# Description
There is a race condition in Dokan FUSE between opening and closing a file that may lead to a use after free.

Assume a file a.txt is already open with handle h1. Then file a.txt is opened again with handle h2 and at the same time handle h1 is closed on another thread. Assume further that the thread closing the file executes up until (but not including) [line 1004 in fusemain.cpp](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L1004). Now the thread opening the file executes up until [line 965 in fusemain.cpp](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L965). Note that since the closing thread has already left the critical section on [line 998](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L998), the opening thread can enter the critical section on [line 948](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L948). Moreover, observe that the closing thread has not yet removed the `impl_file_lock` instance from the `file_locks` map, the opening thread finds the `impl_file_lock` instance on [line 949](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L949) and thus takes the following if-branch. Now we let the thread closing the file proceed by removing the file from the locks map ([line 1004](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L1004)). On [line 1009](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L1009) the same `impl_file_lock` instance that was found before by the thread opening the file, is now found again and is subsequently deleted on [line 1012](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L1012). If the opening thread now proceeds by adding the file to the `impl_file_lock` instance ([line 972](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L972)) we have a use after free of the member `first` on [line 980](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L980).

I think this problem can be fixed by moving the call to `LeaveCriticalSection` on [line 959](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L959) before both [lines 962](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L962) and [974](https://github.com/dokan-dev/dokany/blob/3361dbf7e681d45d71753b2931967e0bd700e492/dokan_fuse/src/fusemain.cpp#L974).

**Comment by Liryna** [comment]
Hi @mistersandman ,

Sorry for the delay, thank you for the analyze and describing in detail how the issue could happen 👍 

I understood the situation that you describe, I just don't understand the possible fix you propose (the number of lines seems to be wrong or I didn’t get).
Can you check if lines are correct or even provide a patch for the issue ?

**Comment by mistersandman** [comment]
Hi @Liryna 

I'm sorry, I actually meant to attach a patch but somehow I forgot to do so. Here you go :-)
[patch_race_cond.txt](https://github.com/dokan-dev/dokany/files/504360/patch_race_cond.txt)

**Comment by Liryna** [comment]
Thank you @mistersandman for the patch ! very appreciate 👍 

Did you faced the issue or did you review the code ?

I have added it https://github.com/dokan-dev/dokany/commit/4f1d82fe80b574d81dabb59debcfbf1972e4e4b1

**Comment by mistersandman** [comment]
Fortunately you caught the mistake of the missing curly braces for the if-branch in my patch, sorry for that!

I actually faced the issue (segfault because of use after free), but I could not reliably debug, because it is very hard to make the kernel-mode driver call into user-mode on different threads on purpose. So after I knew what access triggered the segfault, I reviewed the code and finally found the issue.
